Minseo

superozing opened the pull request.

superozing committed:
#8 파일 위치 조정, 슈도 코드 작성

superozing committed:
원상 복귀...
다시 되돌렸습니당.

superozing committed:
#8 User 클래스, initUser() 함수
initUser()함수 - 회원 등록
 1) 회원 정보 입력 받기
 2) 회원 생성시 중복 체크
   만약 find(이름)의 반환 값이 end iterator라면 이름을 수정할 것을 요청
 3) 추가 입력 여부 묻기

superozing committed:
#8 분할 구현

jeontigger commented:
동명이인은 같은인물이라고 볼 수 없음
find에 실패 => 중복 없음 => 근데 중복이 없을 때 다시 입력해달라 함
분할 굿굿

superozing committed:
#8 유저 정보 출력 함수 구현, main 함수 정리
2. 회원 출력
 2-1) 이름순으로 정렬해서 출력 (역순 선택 가능하도록)
 2-2) 성별에 따라 필터링해서 출력
 2-3) 이름으로 정보 검색

조건 체크 오류 수정 완료
main을 인호 코드처럼 간단하게 바꿔봄.
나이대(10대, 20대 등)에 따라 출력하는 기능은 구현 할 예정.

superozing committed:
#8 UTF-8 변환

jeontigger commented on Minseo/UserFunc.cpp:
```diff
@@ -90,29 +90,29 @@ void outputSort_Name(int isReverse)
-				cout << "����" << endl;
-			cout << "�׷�: " << data.m_Group << endl;
+				cout << "여성" << endl;
+			cout << "그룹: " << data.m_Group << endl;
```
계속 반복되는 형태 => 함수화

superozing replied:
@jeontigger 
main을 줄인 이유는 전에 스터디 때 코드 리뷰 했을 때 메인 세 줄로 줄인 코드가 너무 마음에 들었어
runUserInfoSystem 이름에 run 들어간 것도 영향이 있음

runUserInfoSystem 안에서 while을 돌리지 않고 밖으로 빼놓고, 반환 타입을 int로 설정한 이유는 아마 없을 테지만 나중에 다른 분기가 생길 경우 그에 대한 예외 처리를 하고 싶어서 그랬어.  
전에 수업에서 인자의 반환 타입을 errno_t로 받는 함수가 있는 것을 봤었는데 오류에 따라서 반환을 다르게 돌려주잖아. 
그거 생각나서 했는데 내가 받는 값은 오류를 반환 받는 게 아니니까 일단 int로 받았어.

Hint) 파라미터가 꼭 필요할까?
만들어 놓은 enum 값으로 넣으면 훨씬 좋겠구나!!! 바로 수정해볼게 <- 만약에 아니면 수업 끝나고 알려줄래?

jeontigger replied:
@superozing 
만약 너가 만든 시스템을 다른 프로그래머가 접목시킬 수 있도록 생각할 때, 헤더를 선언하고 어떤 행동들을 취해야할지 생각해보는게 좋아. 그래서 나는 내 유저가 프로그램 헤더를 선언하고 객체 생성 후 run만 해주면 돌아가게 만든것이고, 너의 프로그램은 현재 시점에서는 루프도 겉에 쳐줘야하고, 그 루프 분기조건까지 설정해줘야 하는 거겠지.
메인을 3줄로 만드는게 중요하다기보단, "너의 프로그램 유저가 어떤 방식으로 사용하게 될까?" 를 생각하는게 좋아!

그리고 파라미터가 필요한지에 대한 이야기는 내 생각에outputSort_name의 설계가 두 가지 방법이 있다고 생각했던거야. 1. outputSort_name에 reverse인지 아닌지를 파라미터로 넘겨주어 case 두개에서 같은 함수를 호출하지만 파라미터로 구분하는 방법과
2. outputSort_name_forward()와 outputSort_name_reverse()로 두개로 쪼개서 case안에 따로 넣어주는 방법 - outputSort_name안에 들어가자마자 분기가 이뤄져서 전혀 다른 함수의 역할을 하고 있더라고!
- 1번은 이해한 것 같아!
- 2번으로 하라는 이야기가 아니고, 1번뿐만아니라 2번 구조도 있다고 알려주고싶었어
- 너 스스로 확장성과 변동을 고려해서 너의 구조설계를 해봐!

- int 반환은 굿굿! 스스로 의도가 뭔지 알고있는지 확인 완료!

jeontigger commented on Minseo/UserFunc.cpp:
```diff
@@ -229,16 +229,16 @@ int RunUserInfoSystem()
-	cout << "3. ȸ�� ���� ����" << endl;
-	cout << "4. ������" << endl << endl << endl;
-	cout << "���ϴ� ����� �Է��ϼ���: ";
+	cout << "+++++회원정보 입출력 프로그램+++++" << endl;
+	cout << "---현재 입력된 유저의 수: " << userMap.size() << endl;
+	cout << "1. 회원 등록" << endl;
+	cout << "2. 회원 출력" << endl;
+	cout << "3. 회원 정보 수정" << endl;
+	cout << "4. 끝내기" << endl << endl << endl;
+	cout << "원하는 기능을 입력하세요: ";
```
USER_FUNC 도 생성했는데, 번호 출력도 enum 사용하면 코드 고칠 일이 down!

jeontigger reviewed (commented):
푸시를 자주 해줘서 디테일하게 볼게!

1. 메인을 줄인 이유는?
- 본인의 행동판단근거가 필요
2. runUserInfoSystem함수 리턴 자료형이 int인 이유는?
3. initUser의 함수명 재고려 회원 등록과 의미가 상이한듯
4. outputSort_Name()의 파라미터 설계가 outputSort_Name의 케이스 번호가 달라지면 동작 안할 가능성↑, 그걸 인지하고 있는거 자체가 뇌 메모리 코스트니까 피할 수 있는설계로 재고려하는게 좋을듯! outputSort_Sex()도 마찬가지! 지금 여긴 outputSort_Sex의 케이스 번호 뿐만아니라 성별의 번호가 바뀔 가능성까지 고려해야함!
- Hint) 파라미터가 꼭 필요할까?
- User 출력함수 구현의 의미가 여기서 돋보일듯!

물음표로 끝난 질문 답은 여기 밑으로 커맨트 달아주고
하나라도 왜 해야하는지 이해가 안간다면 같이 얘기해보자

superozing committed:
#8 함수 분할, 유저 탐색 후 수정, 그룹 별 탐색, 나이대 별 탐색 추가, 전체 초기화, main 정리
반복되는 기능을 함수로 묶어버리니까 확실히 코드가 깔끔해짐. 특히 탐색 쪽 함수가 5줄로 끝나는게 기분이 좋더라
윈도우 api 코드 보니까 if문 안에 함수를 넣어서 실행시키고 반환 타입으로 if 구문 조건문을 쓰는 것을 보고 나도 저런 방식을 써보고 싶었음.
함수를 인자를 받지 않고 나누어서 쓰니까 보기도 좋고 쓰기도 좋은 느낌
앞으로도 꼭 인자가 필요한 상황이 아니라면 이런 식으로 써야겠다고 생각함
main도 함수 하나로 깔끔하게 정리했고
기능들 전부 추가 완료

superozing committed:
#8 함수 명 변경
initUser -> registrationUser

jeontigger reviewed (commented):
1. UserInfoSystem함수와 RunUserInfoSystem함수를 분리한 이유는? 따로 무언가 추가될 수 있다고 본건가?
2. outputSort_[]함수들 Name을 제외하고 Sort라기보단 Filter나 추출에 가깝다고 생각함!
- 내가 이런 네이밍 관련 이야기해서 수정할 때 혹시 vs 기능을 쓰고 있나? 일일이 바꾸나?
3. RunUserInfoSystem의 출력이 기능과 일치하지 않는듯!

superozing commented:
1. 전에 나중에 추가할 다른 예외 처리가 있을 경우를 대비해서 while문을 함수 밖으로 빼놓고 반환 타입을 받아서 while문 반복을 체크했는데 사용자 입장에서 함수 하나 사용하면 끝나야 할 정도로 명확한 기능을 가지고 있어야 한다고 했어서 아예 while문 자체를 실행하는 함수를 만들었어. while문을 기존 함수에 넣는 것이 약간 아까워서...
2. 오케이 고쳐볼게. 고칠 때 보통은 vs기능 중 컨트롤 R +컨트롤 R 써서 고치고 있어
3. 이해하지 못했는데 조금 더 자세한 설명 가능할까?
@jeontigger

jeontigger commented:
2. 일일이 바꾸고 있지만 않으면 괜찮지! 난 vscode로 하다보니 F2버튼을 누르고, 난 vs에서는 ctrl+shift+f버튼 썼었어!
3. 내가 착각하고 말한 것 같아 안고쳐도 될듯!
